pkg/argocd/update.go

```diff
@@ -286,20 +286,32 @@ func UpdateApplication(updateConf *UpdateConfiguration, state *SyncIterationStat
 		}
 
 		if needsUpdate(updateableImage, applicationImage, latest) {
+			appImageWithTag := applicationImage.WithTag(latest)
+			appImageWithTagFull := appImageWithTag.GetFullNameWithTag()
+
+			// Check if new image is already set in Application Spec when write back is set to argocd
+			// and compare with new image
+			appImageSpec, err := getAppImage(&updateConf.UpdateApp.Application, appImageWithTag)
+			if err != nil {
+				continue
+			}
+			if appImageSpec == appImageWithTagFull {
+				imgCtx.Infof("New image %s already set in spec", appImageWithTagFull)
+				continue
+			}
 
-			imgCtx.Infof("Setting new image to %s", applicationImage.WithTag(latest).GetFullNameWithTag())
 			needUpdate = true
+			imgCtx.Infof("Setting new image to %s", appImageWithTagFull)
 
-			err = setAppImage(&updateConf.UpdateApp.Application, applicationImage.WithTag(latest))
+			err = setAppImage(&updateConf.UpdateApp.Application, appImageWithTag)
 
 			if err != nil {
 				imgCtx.Errorf("Error while trying to update image: %v", err)
 				result.NumErrors += 1
 				continue
 			} else {
-				containerImageNew := applicationImage.WithTag(latest)
-				imgCtx.Infof("Successfully updated image '%s' to '%s', but pending spec update (dry run=%v)", updateableImage.GetFullNameWithTag(), containerImageNew.GetFullNameWithTag(), updateConf.DryRun)
-				changeList = append(changeList, ChangeEntry{containerImageNew, updateableImage.ImageTag, containerImageNew.ImageTag})
+				imgCtx.Infof("Successfully updated image '%s' to '%s', but pending spec update (dry run=%v)", updateableImage.GetFullNameWithTag(), appImageWithTag.GetFullNameWithTag(), updateConf.DryRun)
```

Comment:
looks like `appImageWithTag.GetFullNameWithTag()` in the above log line can also be replaced with the variable `appImageWithTagFull`. Can you rebase and fix it if you agree, to have CI run with the latest?

Reply:
you are completely right. fixed!